include/deal.II/base/tensor.h

```diff
@@ -19,741 +19,213 @@
-    subtensor[i] *= s;
-  return *this;
-}
-
-
-
-template <int rank_, int dim, typename Number>
-inline
-Tensor<rank_,dim,Number> &
-Tensor<rank_,dim,Number>::operator /= (const Number s)
-{
-  for (unsigned int i=0; i<dim; ++i)
-    subtensor[i] /= s;
-  return *this;
-}
-
-
-
-template <int rank_, int dim, typename Number>
-inline
-Tensor<rank_,dim,Number>
-Tensor<rank_,dim,Number>::operator + (const Tensor<rank_,dim,Number> &t) const
-{
-  Tensor<rank_,dim,Number> tmp(*this);
-
-  for (unsigned int i=0; i<dim; ++i)
-    tmp.subtensor[i] += t.subtensor[i];
-
-  return tmp;
-}
 
+/* ----------------- Non-member functions operating on tensors. ------------ */
 
+/**
+ * @name Output functions for Tensor objects
+ */
+//@{
 
+/**
+ * Output operator for tensors. Print the elements consecutively, with a space
+ * in between, two spaces between rank 1 subtensors, three between rank 2 and
+ * so on.
+ *
+ * @relates Tensor
+ */
 template <int rank_, int dim, typename Number>
 inline
-Tensor<rank_,dim,Number>
-Tensor<rank_,dim,Number>::operator - (const Tensor<rank_,dim,Number> &t) const
+std::ostream &operator << (std::ostream &out, const Tensor<rank_,dim,Number> &p)
 {
-  Tensor<rank_,dim,Number> tmp(*this);
-
-  for (unsigned int i=0; i<dim; ++i)
-    tmp.subtensor[i] -= t.subtensor[i];
+  for (unsigned int i=0; i<dim-1; ++i)
+    out << p[i] << ' ';
+  out << p[dim-1];
 
-  return tmp;
+  return out;
 }
 
 
-
-template <int rank_, int dim, typename Number>
+/**
+ * Output operator for tensors and dimension 1. This is implemented
+ * specialized from the general template in order to avoid a compiler
+ * warning that the loop is empty.
+ *
+ * @relates Tensor
+ */
+template <int rank_>
 inline
-Tensor<rank_,dim,Number>
-Tensor<rank_,dim,Number>::operator - () const
+std::ostream &operator << (std::ostream &out, const Tensor<rank_,1> &p)
 {
-  Tensor<rank_,dim,Number> tmp;
-
-  for (unsigned int i=0; i<dim; ++i)
-    tmp.subtensor[i] = -subtensor[i];
+  out << p[0];
 
-  return tmp;
+  return out;
 }
 
 
-
-template <int rank_, int dim, typename Number>
+/**
+ * Output operator for tensors of rank 0. Since such tensors are scalars, we
+ * simply print this one value.
+ *
+ * @relates Tensor<0,dim,Number>
+ */
+template <int dim, typename Number>
 inline
-typename Tensor<rank_,dim,Number>::real_type
-Tensor<rank_,dim,Number>::norm () const
+std::ostream &operator << (std::ostream &out, const Tensor<0,dim,Number> &p)
 {
-  return std::sqrt (norm_square());
+  out << static_cast<Number>(p);
+  return out;
 }
 
 
-
-template <int rank_, int dim, typename Number>
+/**
+ * Output operator for tensors of rank 1. Print the elements consecutively,
+ * with a space in between.
+ *
+ * @relates Tensor<1,dim,Number>
+ */
+template <int dim, typename Number>
 inline
-typename Tensor<rank_,dim,Number>::real_type
-Tensor<rank_,dim,Number>::norm_square () const
+std::ostream &operator << (std::ostream &out, const Tensor<1,dim,Number> &p)
```

Comment:
do you actually still need this? It looks like it'd be subsumed in the general case for arbitrary rank above.